### PR TITLE
Feature/i18n

### PR DESCRIPTION
**Pull request by mrcruz**
This a a very simple implementation of the way Qt does its translations. It has a script to both extract tagged keys automatically and a script to create a binary file with the translations. 

The main problem that I really would like to improve is in file src/run_tribler.py: I had to shuffle around the imports in the middle of the file because their order does change the semantics of the program. If I import TriblerWindow before installing a translator, the window in created with the default language. That happens because a lot of things are done in constructors. Is there a way to solve this that does not required a huge change?

### CODE REVIEW

**Comment by ichorid** [comment]
ok to test

**Comment by ichorid** [comment]
@mrcruz , welcome and thanks for contributing! :+1:

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
Cool!

> If I import TriblerWindow before installing a translator, the window in created with the default language. That happens because a lot of things are done in constructors. Is there a way to solve this that does not required a huge change?

The easiest and cleanest way is probably to drop the `installTranslator` code into the GUI `__init__.py` file:

https://github.com/Tribler/tribler/blob/devel/src/tribler-gui/tribler_gui/__init__.py

**Comment by qstokkink** [comment]
This is the cleanest I could come up with:

```patch
diff --git a/src/run_tribler.py b/src/run_tribler.py
index e1ff95336..596ecab54 100644
--- a/src/run_tribler.py
+++ b/src/run_tribler.py
@@ -139,14 +139,12 @@ if __name__ == "__main__":
             check_free_space()
 
             from tribler_gui.tribler_app import TriblerApplication
+            from tribler_gui.tribler_window import TriblerWindow
 
             app_name = os.environ.get('TRIBLER_APP_NAME', 'triblerapp')
             app = TriblerApplication(app_name, sys.argv)
-        
-            from tribler_gui.i18n import get_default_system_translator
-            translator = get_default_system_translator()
-            app.installTranslator(translator)
-            
+            app.installTranslator(app.translator)
+
             if app.is_running():
                 for arg in sys.argv[1:]:
                     if os.path.exists(arg) and arg.endswith(".torrent"):
@@ -155,7 +153,6 @@ if __name__ == "__main__":
                         app.send_message(arg)
                 sys.exit(1)
 
-            from tribler_gui.tribler_window import TriblerWindow
             window = TriblerWindow()
             window.setWindowTitle("Tribler")
             app.set_activation_window(window)
diff --git a/src/tribler-gui/tribler_gui/tribler_app.py b/src/tribler-gui/tribler_gui/tribler_app.py
index 9cc12d201..172d97f09 100644
--- a/src/tribler-gui/tribler_gui/tribler_app.py
+++ b/src/tribler-gui/tribler_gui/tribler_app.py
@@ -5,6 +5,7 @@ from PyQt5.QtCore import QCoreApplication, QEvent, Qt
 
 from tribler_core.utilities.unicode import ensure_unicode
 
+from tribler_gui.i18n import get_default_system_translator
 from tribler_gui.code_executor import CodeExecutor
 from tribler_gui.single_application import QtSingleApplication
 
@@ -23,6 +24,7 @@ class TriblerApplication(QtSingleApplication):
         QtSingleApplication.__init__(self, app_name, args)
         self.code_executor = None
         self.messageReceived.connect(self.on_app_message)
+        self.translator = get_default_system_translator()
 
     def on_app_message(self, msg):
         if msg.startswith('file') or msg.startswith('magnet'):
```

**Comment by mrcruz** [comment]
I followed the suggestion by @qstokkink and I would say the only thing missing now is some UI to select the language. I'm trusting the locale.getdefaultlocale() will be reliable even though I do not have any experience with it.

File: src/tribler-gui/tribler_gui/i18n.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+def get_translator_for(locale_id):
+    translator = QTranslator()
+    translator.load(locale_id, "tribler-gui/tribler_gui/i18n")
+    return translator
+
+def get_default_system_translator():
+    defaultLocale = locale.getdefaultlocale()[0]
+    return get_translator_for(defaultLocale)
+
```
Comment (qstokkink) [inline src/tribler-gui/tribler_gui/i18n.py:19]: Could you remove the trailing spaces from this line*?

*This should've been caught by `pylint` with a `C0303`, but apparently it's broken? 🤔

Comment (devos50) [reply src/tribler-gui/tribler_gui/i18n.py:19]: This seems to be an error then in the diff-cover tool, since it triggers when running `pylint` normally.

Comment (qstokkink) [reply src/tribler-gui/tribler_gui/i18n.py:19]: Flake8 could've also caught this I guess, but it is configured for only `--select=E901,E999,F821,F822,F823,I100,I101,I201,I202`.

Comment (ichorid) [reply src/tribler-gui/tribler_gui/i18n.py:19]: Please use `pre-commit` package to run pre-commit checks and lints automatically. We use the Black formatter with the GUI code.

Comment (mrcruz) [reply src/tribler-gui/tribler_gui/i18n.py]: I really tried to read most of the documentation before making the pull request but I completely missed this pre-commit thing. Done that for the files that I have changed now. Should we put this information in a easier to find place?

**Comment by mrcruz** [comment]
I feel the need to create a simple README  to facilitate future translations. Where should I put this? I was thinking in a README file inside de i18n folder. Do we have a standard for this?

File: src/tribler-gui/tribler_gui/i18n.py (hunk 1 above)
Comment (ichorid) [reply src/tribler-gui/tribler_gui/i18n.py]: One can say pre-commit was optional and experimental for some time. `contributing.rst` needs to be updated for sure. Also, pre-commit is only used in the GUI, and there are no checks for Black formatting in the PR tester.

**Comment by qstokkink** [comment]
@mrcruz if you're struggling with rounding off this PR and you have `allow edits from maintainers` enabled, I can help you out with rounding off this PR if you want.

**Comment by mrcruz** [comment]
> @mrcruz if you're struggling with rounding off this PR and you have `allow edits from maintainers` enabled, I can help you out with rounding off this PR if you want.

It is enabled. Feel free to change in any way you want!

**Comment by qstokkink** [comment]
@mrcruz alright 👍 let me polish this up for merge.

**Comment by qstokkink** [comment]
Alright, cleaned up. All tests (should) pass now. 

@mrcruz your e-mail and username were the wrong way around (your username was in the e-mail field and vise versa). I corrected it for this commit, but you'll probably want to fix that locally.

File: src/tribler-gui/tribler_gui/i18n/README.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+This app uses QT multi language feature to provider internationalization. 
+
+## Improving a translation
+
+If you find a mistake on a existing language, a missing translation or any other problem, you can help us. Just find right `<source>` key in the respective TS file and change the value of the `<translation>` key.
+
+## New translations
+
+If you want to translate Tribler to a new languange, you need to create a new TS file. The file is named according to [rfc1766](https://tools.ietf.org/html/rfc1766.html) (pt_BR, for example). You can easy create a new file by adding a new line in the script `extract-messages` and running it.
```
Comment (devos50) [inline src/tribler-gui/tribler_gui/i18n/README.md:11]: typo: languange -> language

Comment (qstokkink) [reply src/tribler-gui/tribler_gui/i18n/README.md]: Fixed 👍

**Comment by devos50** [comment]
Thanks!

**Comment by qstokkink** [comment]
Agreed, very cool.

File: src/tribler-gui/tribler_gui/i18n/README.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# Internationalization
+
+This app uses QT multi language feature to provider internationalization.
```
Comment (linsui) [inline src/tribler-gui/tribler_gui/i18n/README.md:3]: provider: A typo?

Comment (qstokkink) [reply src/tribler-gui/tribler_gui/i18n/README.md:3]: 🤔 indeed. This should also state "**the** QT multi language feature".

Comment (qstokkink) [reply src/tribler-gui/tribler_gui/i18n/README.md:3]: I'll make a quick PR for that, before we forget.

**Comment by ichorid** [comment]
@mrcruz , I've recently [cleaned up and wrapped](https://stackoverflow.com/questions/44286102/minimize-to-tray-on-macosx) all the remaining GUI strings in `tr()`. Would you care to update BR translation?

**Comment by mrcruz** [comment]
I guess you linked the wrong url in your comment. I will work on an update later on this week and create a PR. I'm glad we have more languages already!

**Pull request merged**